dbms/src/DataTypes/DataTypeDateTime.h

```diff
@@ -1,21 +1,35 @@
 #pragma once
 
+#include <Core/Types.h>
 #include <DataTypes/DataTypeNumberBase.h>
 
-
 class DateLUTImpl;
 
 namespace DB
 {
 
+class TimezoneMixin
+{
+public:
+    explicit TimezoneMixin(const String & time_zone_name = "");
+    TimezoneMixin(const TimezoneMixin &) = default;
+
+    const DateLUTImpl & getTimeZone() const { return time_zone; }
+
+protected:
+    bool has_explicit_time_zone;
+    const DateLUTImpl & time_zone;
+    const DateLUTImpl & utc_time_zone;
```

Comment:
Why fields are protected? Is it supposed to inherit this class? (Need comment)

Reply:
Yes, I see it below.
Wouldn't it be better to own this class, but not inherit?

Reply:
well, we can have this as an attribute, but then I'll have to add two identical functions to `DateTime` and `DateTime64`, or add artificial "kinda-base class":

```
template <typename Base>
class DateTimeBase : public Base
{
public:
    DateTimeBase(const String & time_zone_name = "");
    DateTimeBase(const DateTimeBase &);
    const DateLUTImpl & getTimeZone() const { return time_zone; }

protected:
    bool has_explicit_time_zone;
//...
};
```

IMO, mixing is the lesser of three evils here: no code duplication (compared to attribute), less templates (compared to base class).

Reply:
For field case you would need to implement extra `getTimeZone()` function and probably a constructor from `TimezoneMixin`. I think it's still better then inheritance with multiple classes.

But it's ok to use your approach.